pkg/volume/util.go

```diff
@@ -331,3 +332,215 @@ func ChooseZoneForVolume(zones sets.String, pvcName string) string {
 	glog.V(2).Infof("Creating volume for PVC %q; chose zone=%q from zones=%q", pvcName, zone, zoneSlice)
 	return zone
 }
+
+// Zones2Set converts a string containing a comma separated list of zones to set
+func Zones2Set(zonesString string) (sets.String, error) {
+	zonesSlice := strings.Split(zonesString, ",")
+	zonesSet := make(sets.String)
+	for _, zone := range zonesSlice {
+		trimmedZone := strings.TrimSpace(zone)
+		if trimmedZone == "" {
+			return make(sets.String), fmt.Errorf("comma separated list of zones must not contain an empty zone")
+		}
+		zonesSet.Insert(trimmedZone)
+	}
+	return zonesSet, nil
+}
+
+// ValidatePVCSelector validates Selector part of a PVC:
+// - in case there is no Selector the PVC is valid
+// - makes sure that only allowedKeys are present in the Selector matchLabels part
+// - makes sure that only allowedKeys and allowedOperators are present in the Selector matchExpressions part
+// Return value:
+// - (true, nil) means PVC is valid (error == nil) and there is NO Selector OR (NO matchLabels AND NO matchExpressions) (bool == true)
+// - (false, nil) means PVC is valid (error == nil) and there is at least a value in matchLabels or matchExpressions specified (bool == false)
+// - (false, error) means PVC is not valid
+// - (true, error) shall never happen
+func ValidatePVCSelector(pvc *v1.PersistentVolumeClaim) (bool, error) {
+	allowedKeys := map[string]bool{unversioned.LabelZoneFailureDomain: true, unversioned.LabelZoneRegion: true}
+	allowedOperators := map[unversioned.LabelSelectorOperator]bool{unversioned.LabelSelectorOpIn: true, unversioned.LabelSelectorOpNotIn: true}
+	if pvc.Spec.Selector == nil {
+		return true, nil
+	}
+	if len(pvc.Spec.Selector.MatchExpressions) < 1 && len(pvc.Spec.Selector.MatchLabels) < 1 {
+		return true, nil
+	}
+	if len(pvc.Spec.Selector.MatchLabels) > 0 {
+		for label := range pvc.Spec.Selector.MatchLabels {
+			if !allowedKeys[label] {
+				return false, fmt.Errorf("Persistent volume claim (%v), matchLabels part: the %v key is not permitted", pvc.Name, label)
+			}
+		}
+	}
+	for _, expr := range pvc.Spec.Selector.MatchExpressions {
+		if !allowedKeys[expr.Key] {
+			return false, fmt.Errorf("Persistent volume claim (%v), matchExpressions part: the %v key is not permitted", pvc.Name, expr.Key)
+		}
+		if !allowedOperators[expr.Operator] {
+			return false, fmt.Errorf("Persistent volume claim (%v), matchExpressions part: the %v operator is not permitted", pvc.Name, expr.Operator)
+		}
+		if len(expr.Values) < 1 {
+			return false, fmt.Errorf("Persistent volume claim (%v): the %v key, %v operator does not contain any values", pvc.Name, expr.Key, expr.Operator)
+		}
+	}
+	return false, nil
+}
+
+// GetPVCMatchLabel returns:
+// - either (value, nil) for the key from the matchLabels Selector part of the PVC
+// - or ("", error) in case the key is missing in the matchLabels Selector part of the PVC
+func GetPVCMatchLabel(pvc *v1.PersistentVolumeClaim, key string) (string, error) {
```

Comment:
And there are already two different errors, so it feels like an error is right, I agree